src/state/farms/classic/knUpdater.tsx

```diff
@@ -0,0 +1,223 @@
+import { BigNumber } from '@ethersproject/bignumber'
+import { Token, WETH } from '@kyberswap/ks-sdk-core'
+import { parseUnits } from 'ethers/lib/utils'
+import { useEffect } from 'react'
+import { ClassicFarmKN, useLazyGetFarmClassicQuery } from 'services/knprotocol'
+
+import FAIRLAUNCH_V2_ABI from 'constants/abis/fairlaunch-v2.json'
+import FAIRLAUNCH_V3_ABI from 'constants/abis/fairlaunch-v3.json'
+import FAIRLAUNCH_ABI from 'constants/abis/fairlaunch.json'
+import { AbortedError, ZERO_ADDRESS } from 'constants/index'
+import { isEVM } from 'constants/networks'
+import { useActiveWeb3React } from 'hooks'
+import { useKyberSwapConfig } from 'state/application/hooks'
+import { setFarmsData, setLoading, setYieldPoolsError } from 'state/farms/classic/actions'
+import { FairLaunchVersion, Farm, FarmV1, FarmV2 } from 'state/farms/classic/types'
+import { useAppDispatch } from 'state/hooks'
+import { getContractForReading } from 'utils/getContract'
+import { parseFraction, toString } from 'utils/numbers'
+
+const KNUpdater = ({ isInterval = true }: { isInterval?: boolean }) => {
+  const dispatch = useAppDispatch()
+  const { chainId, account } = useActiveWeb3React()
+  const { isEnableKNProtocol, readProvider } = useKyberSwapConfig()
+  const [fetchFarmKN] = useLazyGetFarmClassicQuery()
+
+  useEffect(() => {
+    const abortController = new AbortController()
+
+    async function getListFarmsKN(): Promise<Farm[]> {
+      try {
+        if (!isEVM(chainId)) return []
+        const farmsKN: ClassicFarmKN[] | undefined = (await fetchFarmKN(chainId)).data?.data.farmPools
+        if (!farmsKN) return []
+
+        const mapping = await Promise.all(
+          farmsKN.map(async farmPool => {
+            const [fairLaunchAddress] = farmPool.id.split('_')
+            const pid = Number(farmPool.pid)
+            const id = farmPool.pool.id
+            // const version =
+            //   farmPool.version === 1
+            //     ? FairLaunchVersion.V1
+            //     : farmPool.version === 2
+            //     ? FairLaunchVersion.V2
+            //     : FairLaunchVersion.V3
+            // todo namgold: subgraph issue, revert this later
+            const version: FairLaunchVersion =
+              Number(farmPool.start) < 1_000_000_000 ? FairLaunchVersion.V1 : FairLaunchVersion.V2
```

Comment:
oops. v3 can't be used here

Reply:
when revert TODO?

Reply:
Definitely, this only temporary solution and not meant to be used after subgraph is fixed. If v3 is integrating when subgraph still issue, we need to find a way to tell the different from v3 to v1 & v2 & update this place.